File: src/02.js
import { getInputLines } from './common/common.js'

function solve_02(commands) {
  let hpos = 0, depth_1 = 0, depth_2 = 0, aim = 0
  for (const command of commands) {
    switch (command.direction) {
      case "forward":
        hpos += command.value
        depth_2 += aim * command.value
        break;
      case "up":
        depth_1 -= command.value
        aim -= command.value
        break;
      case "down":
        depth_1 += command.value
        aim += command.value
        break;
    }
  }
  return [hpos * depth_1, hpos * depth_2]
}

function parse(line) {
  const parts = line.split(' ')
  return {direction: parts[0], value: Number(parts[1])}
}

const commands = getInputLines().map(parse)
const [result_1, result_2] = solve_02(commands)
console.log(result_1)
console.log(result_2)

File: src/02.rb
require './src/common/common.rb'

HUGE = 9999

def solve(commands)
  hpos = depth_1 = depth_2 = aim = 0
  commands.each do |command|
    case command.direction
    when 'forward'
      hpos += command.value
      depth_2 += aim * command.value
    when 'up'
      depth_1 -= command.value
      aim -= command.value
    when 'down'
      depth_1 += command.value
      aim += command.value
    end
  end
  return hpos * depth_1, hpos * depth_2
end

Command = Struct.new(:direction, :value)

parse = lambda do |line|
  parts = line.split
  Command.new(parts[0], parts[1].to_i)
end

if __FILE__ == $0
  commands = get_parsed_lines(parse)
  result_1, result_2 = solve(commands)
  puts result_1
  puts result_2
end
